### ISSUE TITLE
LB boundary forces wrong

### ISSUE THREAD

**Issue by KaiSzuttor** [espressomd/espresso#2624]
The test is very sensitive on parameter changes which makes me doubt that it's correctly testing what it should.

**Comment by mkuron** [comment]
Confirmed. I ported that test to Walberla to test the boundary force and it didn't even give the same results for the same parameters. See https://github.com/espressomd/espresso/issues/2540

**Comment by KaiSzuttor** [comment]
``` python
import itertools
import numpy as np

import espressomd
import espressomd.lb
import espressomd.lbboundaries
import espressomd.shapes

AGRID = .5
DENS = 1.5
VISC = 2.3
TIME_STEP = 0.1

LB_PARAMS = {'agrid': AGRID,
             'dens': DENS,
             'visc': VISC,
             'tau': TIME_STEP}

SPHERE_RADIUS = 5.5
BOUNDARY_VELOCITY = [2.3, 0.0, 0.0]

system = espressomd.System(box_l=[3*SPHERE_RADIUS]*3)                                                                                  
system.cell_system.skin = 0.5 * AGRID
system.time_step = TIME_STEP

lb_fluid = espressomd.lb.LBFluidGPU(**LB_PARAMS)
system.actors.add(lb_fluid)

# Add spherical LB boundary
sphere = espressomd.lbboundaries.LBBoundary(shape=espressomd.shapes.Sphere(radius=SPHERE_RADIUS, center=system.box_l*0.5, direction=1))
system.lbboundaries.add(sphere)

wall_shapes = [espressomd.shapes.Wall(normal=[0, 0, 1], dist = 0.75 * AGRID),
               espressomd.shapes.Wall(normal=[0, 0, -1], dist = -(system.box_l[2] - 0.75 * AGRID)),
               espressomd.shapes.Wall(normal=[0, 1, 0], dist = 0.75 * AGRID),
               espressomd.shapes.Wall(normal=[0, -1, 0], dist = -(system.box_l[1] - 0.75 * AGRID))]

for ws in wall_shapes:
    system.lbboundaries.add(espressomd.lbboundaries.LBBoundary(shape=ws, velocity=BOUNDARY_VELOCITY))


system.integrator.run(1000)
#number_of_LB_nodes = np.array(system.box_l / AGRID, dtype=int)
#fluid_node_indices = itertools.product(range(number_of_LB_nodes[0]),
#                                       range(number_of_LB_nodes[1]),
#                                       range(number_of_LB_nodes[2]))
#for index in fluid_node_indices:
#    if (lb_fluid[index].boundary == 0):
#        print(lb_fluid[index].velocity[0])


stokes_force = 6.0 * np.pi * VISC * SPHERE_RADIUS * BOUNDARY_VELOCITY[0]
print(sphere.get_force()[0], stokes_force)
```

Seems like there is an error in the force calculation for the boundaries.

**Comment by fweik** [comment]
The boundary forces code is highly suspicious, that probably needs to be rewritten or removed.

**Comment by mkuron** [comment]
In your test case with a sphere confined between walls, you don't exactly expect Stokes law -- the sphere should be moving a bit slower. (In your case but with only two walls, you'd expect around 30% slower per figure 6a of https://arxiv.org/abs/1503.02681. But since you have four walls, that might be even less).

**Comment by RudolfWeeber** [comment]
Your box is too small, I think. 
For CPU, the results get much less suspicious, if you go from 1.5 diameters to 4-6 diameters.
For GPU, the results seem wrong for all cases

**Comment by KaiSzuttor** [comment]
Still it does not give the correct force and the trend is wrong (the force gets smaller for increasing box size).

**Comment by jainka** [comment]
I must admit that I am quite confused with this. Can you please confirm if the LB parameters that you set in your script are within the Mach number limits of the LBM? I did a quick parameters analysis in _Musubi_ and I found that for this setup (this high velocity and viscosity) you need a time step of at least **0.002**.

With that said, I ran your script upon successively halving the time step. The force computed from **get_force** method indeed seems to converge to the analytical stokes_force with each such step. 

1. dt = 0.1 (152.66885375976562, 21.460219416671872)
2. dt = 0.05 (152.4984130859375, 21.460219416671872)
3. dt = 0.025 (132.29556274414062, 21.460219416671872)
4. dt = 0.0125 (79.73605346679688, 21.460219416671872)
5. dt = 0.006125 (26.244016647338867, 21.460219416671872)

This is a classic LB convergence behavior. 
I am continuing this analysis and would be glad if you can meanwhile confirm my confusion.

**Comment by jainka** [comment]
I forgot to mention that I also changed the high velocity of 2.3 to 0.09. Will try to see where the problem is. From intuition, it seems like some buggy multiplying factor in get_forces.

**Comment by jainka** [comment]
hmmmm.... when the simulation is run much longer the error seems similar... will do more rigorous parametric variations now...

**Comment by mkuron** [comment]
Could you please try implementing this simpler test case in Espresso: https://i10git.cs.fau.de/walberla/walberla/blob/c5f81ce2bc58de53debc803186a96677cb86a3cf/tests/lbm/boundary/BoundaryForceCouette.cpp . It's the force on the walls due to Couette flow. This has way fewer potential issues (finite box size, discretization, etc.) than the sphere.

**Comment by jainka** [comment]
there are even more simpler test cases, which would be better to implement. I need some time to think and understand whats going on...

**Comment by RudolfWeeber** [comment]
Fixed by #2633
